### ISSUE TITLE
problems with static resources with HTTP/2 enabled /FISH-61

### ISSUE THREAD

**Issue by eduarddrenth** [payara/Payara#2625]
# Description #
----------

I have a simple jsf 2 application with resources under javax.faces.resource. On and of there is a problem with downloading these, it either takes very long or fails, messing up the layout. Pressing F5 several times will finaly result in a correct page.

## Expected Outcome

no problems with downloading static resources

## Current Outcome

non responsive application with layout messed up

In the log I see:

- JSF1064: Unable to find or serve resource
- org.glassfish.grizzly.http.io.OutputBuffer.blockAfterWriteIfNeeded(OutputBuffer.java:1023)
- Unable to push resource identified by path
- org.glassfish.grizzly.nio.NIOConnection.assertOpen(NIOConnection.java:464)

## Steps to reproduce (Only for bug reports) 

Perhaps I'll spend time on reproducing, but I've just spent a few days already trying to solve this. My decision is to stick with 4.1.2.181 which is very stable.

## Environment ##

- **Payara Version**: 5.181
- **Edition**: full
- **JDK Version**: 1.8.0_162-8u162-b12-0ubuntu0.16.04.2-b12
- **Operating System**: inux 4.4.0-119-generic
- **Database**: MySQL

**Comment by phillipross** [comment]
I've seen this too, and it's blocking us from moving to Payara5, but now that I've seen I'm not the only one, I think it may be worth the time to try to create a reproducer.  In my case, the problem doesn't occur on every single request so it's a little difficult to reproduce.

**Comment by lprimak** [comment]
This looks like an issue with server push in HTTP/2.  Can you confirm this?

**Comment by lprimak** [comment]
Also looks like a Grizzly issue

**Comment by phillipross** [comment]
I don't know about server push, but there's definitely lots of mention of grizzly in the stack traces.  I'll try to setup another deployment of my project with payara 5.181 and gather some stack traces for reference.

**Comment by eduarddrenth** [comment]
I think I can manage to provide a reproducer of https://bitbucket.org/fryske-akademy/superlemma-list:

- payara preparation script (already there I think)
- mysql dump
- war (or just create that using mvn verify)
- it uses filerealm and needs a group Superlemma

Maybe tomorrow, but I think later....

**Comment by smillidge** [comment]
Try disabling http/2 on the listener and see if the issue goes away.

**Comment by lprimak** [comment]
Great idea

**Comment by eduarddrenth** [comment]
just tested, problem gone

**Comment by smillidge** [comment]
Ok looks like an issue with server push in http/2 as Kenny suggests.

**Comment by smillidge** [comment]
Sorry Lenny my phone called you Kenny.
We also have fixed some issues recently in Grizzly related to http/2 protocol handling so a simple test case would be really useful.

**Comment by phillipross** [comment]
Disabling HTTP/2 on the listeners worked for me too... I receive no errors after that.

**Comment by eduarddrenth** [comment]
shall I close this issue?

**Comment by lprimak** [comment]
> Sorry Lenny my phone called you Kenny.
I get that a lot 🥇

**Comment by eduarddrenth** [comment]
Reproducer

- install mysql
- install database from [sup.sql.gz](https://github.com/payara/Payara/files/1910021/sup.sql.gz)
- unzip payara
- put [mysql-connector-java-6.0.6.jar.gz](https://github.com/payara/Payara/files/1910022/mysql-connector-java-6.0.6.jar.gz) in glassfish/domains/domain1/lib
- download init_glassfish.sh and asadmin_commands.txt from https://bitbucket.org/fryske-akademy/superlemma-list
- adapt host and user in asadmin_commands.txt
- review and execute shellscript
- create a user in file realm with a group Superlemma
- clone https://bitbucket.org/fryske-akademy/superlemma-list
- mvn verify
- deploy war from superlemma-list-gui
- click around in the app

**Comment by MattGill98** [comment]
Hi @eduarddrenth,

The scope of that example is rather large, and I'm having trouble trying to narrow down the issue. Are you able to produce a [concise](http://sscce.org/) reproducer?

Kind regards,

Matt

**Comment by eduarddrenth** [comment]
Hi, I am aware...I'll have to find some time somewhere....

**Comment by eduarddrenth** [comment]
Just tried fresh quarterly, layout isn't messed up anymore with http 2 on, but sometimes a page remains blank, after F5 I see the page. without http 2 all is well. See if I can give you a mvce anywhere soon

**Comment by eduarddrenth** [comment]
I see in the log:
```
Warning:   JSF1064: Unable to find or serve resource, layout/layout.js, from library, primefaces.
Warning:   java.io.IOException: Connection is closed
	at org.glassfish.grizzly.nio.NIOConnection.assertOpen(NIOConnection.java:440)
```
and later:
```
Severe:   Unable to push resource identified by path [/standertwurdlist-gui/faces/javax.faces.resource/jsf.js?ln=javax.faces]
Warning:   JSF1064: Unable to find or serve resource, hotkey/hotkey.js, from library, primefaces.
Severe:   Unable to push resource identified by path [/standertwurdlist-gui/faces/javax.faces.resource/datascroller/images/loading.gif?ln=primefaces&v=6.2.5]
Severe:   Session is closed
org.glassfish.grizzly.http2.Http2StreamException streamId=64 errorCode=REFUSED_STREAM: Session is closed
	at org.glassfish.grizzly.http2.Http2Session.openStream(Http2Session.java:1087)
```

**Comment by eduarddrenth** [comment]
maybe this helps: https://github.com/square/okhttp/issues/2506

**Comment by eduarddrenth** [comment]
enabling http2, but disabling push solves the problem

**Comment by eduarddrenth** [comment]
I'll see what I can do, the app isn't realy that complicated, so I should be able to create a simple reproducer for you....may some time though.

**Comment by eduarddrenth** [comment]
For now we will not do http2, no functional need, apache not quite ready yet, a bit too early.

I'll still see if I can find some time to provide a reproducer.

**Comment by MattGill98** [comment]
Hi @eduarddrenth,

There's [an issue](https://forum.primefaces.org/viewtopic.php?t=56261) raised with PrimeFaces that may be related. Are you able to reproduce the issue without using PrimeFaces?

Kind regards,

Matt

**Comment by eduarddrenth** [comment]
Thanks, that looks very similar indeed... From the discussion I understand there is no clarity whether it is a pf or a payara issue. I'm on holiday now. Reproducing without using pf is I suspect more of a hassle for me than providing a standalone reproducer....

**Comment by lprimak** [comment]
Assigned ```PAYARA-2890``` issue.  This is reproducible

**Comment by eduarddrenth** [comment]
Do I understand correctly that you do not need a reproducer from me anymore?

**Comment by lprimak** [comment]
Correct.  I can reproduce it

**Comment by smillidge** [comment]
We've reproduced it using the PrimeFaces show case.

**Comment by lprimak** [comment]
Can you check if the current trunk has the same issue? I am no longer able to reproduce it since the latest updates:

https://www.payara.fish/upstream_builds

**Comment by phillipross** [comment]
I just tested and it does continue happening for me

**Comment by lprimak** [comment]
Strange... I’ll try again with 5.182 and report back

**Comment by lprimak** [comment]
Hmmm, I can't reproduce the behavior on 5.182 with the same tests.
@phillipross how are you reproducing this? FYI? Do you have any simpler method than clicking around in PrimeFaces showcase (that's what I am doing)?
Thanks

**Comment by phillipross** [comment]
I've been testing with a nontrivial production app (not redistributable) but now that I see that it's been isolated to primefaces, I'll spend some time today to assemble a reproducer.

**Comment by lprimak** [comment]
This is what worked for me in the past, but doesn't reproduce the issue anymore:

```
git clone git@github.com:primefaces/showcase.git
cd showcase
git checkout 6_1
mvn clean package
asadmin deploy target/showcase-6.1.war
```

and click around the app, but most likely Menu->Menu->Non-Ajax Delete

**Comment by phillipross** [comment]
I'm trying it now with the showcase but I'm getting a lot of other errors, and it seems the Multimedia->GraphicImage isn't working.  It doesn't show the images, but it doesn't show any exceptions in the log either.  I'm assuming this is a problem with the showcase though as I've run into this before.  I know our app easily reproduces the problem when using the p:graphicImage component, so I'll try to reproduce it with that.

**Comment by eduarddrenth** [comment]
In my apps I don't use p:graphicImage, but I do have the problem, I can easily reproduce it locally. I can try my app on the latest payara trunk, which branch should I use?

**Comment by phillipross** [comment]
I'm currently testing with Payara Server (full) downloaded from https://www.payara.fish/upstream_builds

I'm also seeing that it's not necessarily related to p:graphicImage, but related to things being served out as static resources.  I'm using primefaces ultima theme, and the errors I get are all referring to the static resources that are part of the theme.

```
[2018-08-15T18:49:09.921+0000] [Payara 5.183-SNAPSHOT] [SEVERE] [] [org.glassfish.grizzly.http2.Http2ServerFilter] [tid: _ThreadID=38 _ThreadName=http-thread-pool::http-listener-2(4)] [timeMillis: 1534358949921] [levelValue: 1000] [[
  Unable to push resource identified by path [/javax.faces.resource/images/profile-bg-dark.png.action?ln=ultima-layout]]]
```

**Comment by phillipross** [comment]
@lprimak just a thought... you may not be able to reproduce the problem if the static resources are cached in your browser

**Comment by lprimak** [comment]
Yea I thought about that as well. Used to be reproductable before though. 
Which browser are you using?

**Comment by phillipross** [comment]
I've only tried latest Chrome [currently Version 68.0.3440.106 (Official Build) (64-bit)]
on MacOS [currently 10.13.6]

**Comment by phillipross** [comment]
Wow... with firefox (61.0.2) I get many other errors.  I forgot to also mention... with my tests, i'm running on https port 8181 (default out of the box ssl certs) and the webapp uses a security constraint to force http traffic to https:
```
   <security-constraint>
        <web-resource-collection>
            <web-resource-name>securedapp</web-resource-name>
            <url-pattern>/*</url-pattern>
        </web-resource-collection>
        <user-data-constraint>
            <!--<transport-guarantee>NONE</transport-guarantee>-->    <!-- NONE does not force to SSL -->
            <transport-guarantee>CONFIDENTIAL</transport-guarantee>   <!-- CONFIDENTIAL forces to SSL -->
        </user-data-constraint>
    </security-constraint>
```

**Comment by lprimak** [comment]
The root cause of this is a race condition of some kind.

**Comment by phillipross** [comment]
At this point, I've confirmed that I'm able to easily reproduce ONLY with the https listener.  As of yet, I've not been able to reproduce with the http listener.

And depending on whether I'm using chrome or firefox, I get different errors and stack traces.

In general, requests from chrome give me the Http2ServerFilter errors, while firefox gives me the javax.enterprise.resource.webcontainer.jsf.application errors... both complaining about static resources.  

I'm still working on minimizing the reproducer

**Comment by smillidge** [comment]
Afaik no browser uses HTTP/2 unless it is https.

**Comment by lprimak** [comment]
Ok, I don't know what was happening before, but I can reliably reproduce it again

**Comment by phillipross** [comment]
I discovered the problem is not related to primefaces.  I now have a reproducer that just pushes static resources through JSF's resources mechanism to cause the problem

source: https://bitbucket.org/luminositylabs/payara-2625

artifact: https://oss.sonatype.org/service/local/repositories/snapshots/content/co/luminositylabs/oss/reproducer/payara-2625/web1/0.0.1-SNAPSHOT/web1-0.0.1-20180816.051644-1.war

I've found that this reproduces the problem with both chrome and firefox.  While testing, I sometimes found that responses stopped coming back the browser and some time would pass before the errors showed up in the server log.  In these cases, I'd restart the app server and then the responses would be quick again with errors showing up in the log immediately.

**Comment by phillipross** [comment]
I added a non-jsf test case which refers to all static resources in a plain html file and the problem does not occur.  This might help narrow down the problem a little more... maybe a problem with mojarra?

**Comment by lprimak** [comment]
Can you see if your non-JSF reproducer does HTTP/2 server push?  I think there is Grizzly race condition when HTTP/2 server push is used.

JSF/PrimeFaces combination sets up resource loading "correctly" which uses server push extensively with HTTP/2, thus more conducive to reproduce this problem

**Comment by phillipross** [comment]
Nope, the pure static non-JSF page doesn't seem to result in any attempts to use push.

**Comment by lprimak** [comment]
Cool, hence the issue is narrowed down to server push alone

**Comment by eduarddrenth** [comment]
update: issue still exists with 5.183

**Comment by lprimak** [comment]
Yes, I am still working on it

**Comment by marcreichman** [comment]
I'm finding this issue to be present for any static content in my app which has gone from GF 3.1.2, through Payara 4.x versions and to 5.183. The default HTTP2 has definitely demonstrated slow loading all over the place. At first I attributed it to lack of gzip compression (which we use on HTTP/1.1), but it's clear the browser is stuck loading resources for a long time. Disabling HTTP/2 and everything goes back to an incredibly snappy experience.

**Comment by phillipross** [comment]
@marcreichman Yes, you're experiencing the exact symptom of this issue.  Disabling HTTP/2 is the current workaround for now, and they're reportedly still working on the issue.

**Comment by lprimak** [comment]
Yes we are working on this issue. The latest version of Payara has server push disabled by default

**Comment by MattGill98** [comment]
Hi all,

This should be fixed by https://github.com/payara/Payara/pull/3690. Feel free to reopen this issue if you have any further issues.

Cheers,

Matt

**Comment by eduarddrenth** [comment]
congrats! exiting, I'll test in 5191...

**Comment by eduarddrenth** [comment]
Sorry, just tested with fresh installed 5.191, still problems with http2 enabled.

Disabling just push doesn't help, still have to disable http2 completely. Linux ff 65

log:
```
java.io.IOException: Write timeout exceeded when trying to flush the data
        at org.glassfish.grizzly.http.io.OutputBuffer.blockAfterWriteIfNeeded(OutputBuffer.java:999)
        at org.glassfish.grizzly.http.io.OutputBuffer.write(OutputBuffer.java:701)
        at org.apache.catalina.connector.OutputBuffer.writeBytes(OutputBuffer.java:336)
        at org.apache.catalina.connector.OutputBuffer.write(OutputBuffer.java:323)
        at org.apache.catalina.connector.CoyoteOutputStream.write(CoyoteOutputStream.java:154)
        at java.nio.channels.Channels$WritableByteChannelImpl.write(Channels.java:458)
        at com.sun.faces.application.resource.ResourceHandlerImpl.handleResourceRequest(ResourceHandlerImpl.java:318)
        at javax.faces.application.ResourceHandlerWrapper.handleResourceRequest(ResourceHandlerWrapper.java:171)
        at org.primefaces.application.resource.PrimeResourceHandler.handleResourceRequest(PrimeResourceHandler.java:87)
```

**Comment by phillipross** [comment]
I also continue having problems with 5.191... both chrome and firefox exhibit the same behavior.  It appears to be correlated to the number of resources on the page.  With a JSF page that only has 20 resources it pushes them out fine, but on pages with 100 resources it never completes and usually requires a restart of the domain before the listener works properly again.

**Comment by eduarddrenth** [comment]
I have around 20 resources and problems

**Comment by MattGill98** [comment]
Hi @eduarddrenth,

With the initial reproducer I experience no problems, but with the updated reproducer (thanks @phillipross!) I find the mentioned errors. Although 5.191 fixed two of the exception causes, it annoyingly appears that there is another. Thank you for raising this, I've created `PAYARA-3671` to address this.

Cheers,

Matt

**Comment by eduarddrenth** [comment]
right, good luck!

**Comment by eduarddrenth** [comment]
tested with 5.193, problem remains

**Comment by awenger63** [comment]
Hi, I just change server machine and replace GF 4.1 by Payara Server 5.193 #badassfish (build 252) and I have similar issue.

**Comment by awenger63** [comment]
The project have also PF 6.2.10, https (CONFIDENTIAL) pages ...

**Comment by phillipross** [comment]
As of 5.192 the http/2 problems were resolved for my projects, and they haven't occurred with 5.193 but I'll test again to see if maybe there was some regression from 5.192 to 5.193.

I think the problem @awenger63 is having is different from the problems I was having since the error message is much different.  Looking at the stackoverflow article, it seems the error has more to do with the number of resources being served.

**Comment by eduarddrenth** [comment]
I have never seen the problem solved as of yet, but face a different error then @awenger63, write timeout exceeded...

**Comment by awenger63** [comment]
Hi, here one complet Log. It' exist a lot of same error for each resources files images, css, js ...  
@phillipross Error of number of ressources appear after a lot of this first error. I thinks, it's a consequence of the number of http request failed and resend.
Hope this can help.

`Unable to push resource identified by path [/MyApp/javax.faces.resource/myLogo.png]|#]
  org.glassfish.grizzly.http2.Http2StreamException streamId=506 errorCode=REFUSED_STREAM
	at org.glassfish.grizzly.http2.Http2Session.openStream(Http2Session.java:1090)
	at org.glassfish.grizzly.http2.Http2ServerFilter.doPush(Http2ServerFilter.java:1017)
	at org.glassfish.grizzly.http2.Http2ServerFilter.handleEvent(Http2ServerFilter.java:603)
	at org.glassfish.grizzly.filterchain.ExecutorResolver$6.execute(ExecutorResolver.java:70)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:260)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:177)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:109)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:88)
	at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:53)
	at org.glassfish.grizzly.filterchain.FilterChainContext.notifyDownstream(FilterChainContext.java:929)
	at org.glassfish.grizzly.filterchain.FilterChainContext.notifyDownstream(FilterChainContext.java:911)
	at org.apache.catalina.core.ApplicationPushBuilder.push(ApplicationPushBuilder.java:365)
	at com.sun.faces.context.ExternalContextImpl.pushIfPossibleAndNecessary(ExternalContextImpl.java:1187)
	at com.sun.faces.context.ExternalContextImpl.encodeResourceURL(ExternalContextImpl.java:644)
	at javax.faces.context.ExternalContextWrapper.encodeResourceURL(ExternalContextWrapper.java:140)
	at org.primefaces.component.graphicimage.GraphicImageRenderer.getImageSrc(GraphicImageRenderer.java:71)
	at org.primefaces.component.graphicimage.GraphicImageRenderer.encodeEnd(GraphicImageRenderer.java:41)
	at javax.faces.component.UIComponentBase.encodeEnd(UIComponentBase.java:595)
	at com.sun.faces.renderkit.html_basic.HtmlBasicRenderer.encodeRecursive(HtmlBasicRenderer.java:286)
	at com.sun.faces.renderkit.html_basic.OutputLinkRenderer.encodeChildren(OutputLinkRenderer.java:100)
	at javax.faces.component.UIComponentBase.encodeChildren(UIComponentBase.java:566)
	at com.sun.faces.renderkit.html_basic.HtmlBasicRenderer.encodeRecursive(HtmlBasicRenderer.java:278)
	at com.sun.faces.renderkit.html_basic.GroupRenderer.encodeChildren(GroupRenderer.java:90)
	at javax.faces.component.UIComponentBase.encodeChildren(UIComponentBase.java:566)
	at javax.faces.component.UIComponent.encodeAll(UIComponent.java:1647)
	at javax.faces.render.Renderer.encodeChildren(Renderer.java:152)
	at javax.faces.component.UIComponentBase.encodeChildren(UIComponentBase.java:566)
	at javax.faces.component.UIComponent.encodeAll(UIComponent.java:1647)
	at javax.faces.component.UIComponent.encodeAll(UIComponent.java:1650)
	at javax.faces.component.UIComponent.encodeAll(UIComponent.java:1650)
	at com.sun.faces.application.view.FaceletViewHandlingStrategy.renderView(FaceletViewHandlingStrategy.java:468)
	at com.sun.faces.application.view.MultiViewHandler.renderView(MultiViewHandler.java:170)
	at javax.faces.application.ViewHandlerWrapper.renderView(ViewHandlerWrapper.java:132)
	at com.sun.faces.lifecycle.RenderResponsePhase.execute(RenderResponsePhase.java:102)
	at com.sun.faces.lifecycle.Phase.doPhase(Phase.java:76)
	at com.sun.faces.lifecycle.LifecycleImpl.render(LifecycleImpl.java:199)
	at javax.faces.webapp.FacesServlet.executeLifecyle(FacesServlet.java:708)
	at javax.faces.webapp.FacesServlet.service(FacesServlet.java:451)
	at org.apache.catalina.core.StandardWrapper.service(StandardWrapper.java:1628)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:258)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:160)
	at org.apache.catalina.core.StandardPipeline.doInvoke(StandardPipeline.java:755)
	at org.apache.catalina.core.StandardPipeline.invoke(StandardPipeline.java:575)
	at com.sun.enterprise.web.WebPipeline.invoke(WebPipeline.java:99)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:159)
	at org.apache.catalina.connector.CoyoteAdapter.doService(CoyoteAdapter.java:371)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:238)
	at com.sun.enterprise.v3.services.impl.ContainerMapper$HttpHandlerCallable.call(ContainerMapper.java:520)
	at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:217)
	at org.glassfish.grizzly.http.server.HttpHandler.runService(HttpHandler.java:182)
	at org.glassfish.grizzly.http.server.HttpHandler.doHandle(HttpHandler.java:156)
	at org.glassfish.grizzly.http.server.HttpServerFilter.handleRead(HttpServerFilter.java:218)
	at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(ExecutorResolver.java:95)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:260)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:177)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:109)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:88)
	at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:53)
	at org.glassfish.grizzly.http2.Http2Session.sendMessageUpstream(Http2Session.java:1282)
	at org.glassfish.grizzly.http2.Http2Session.sendMessageUpstream(Http2Session.java:1258)
	at org.glassfish.grizzly.http2.Http2ServerFilter$2.run(Http2ServerFilter.java:1056)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:569)
	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:549)
	at java.lang.Thread.run(Thread.java:748)`

**Comment by eduarddrenth** [comment]
On 5.194 I don't see
```
java.io.IOException: Write timeout exceeded when trying to flush the data
        at org.glassfish.grizzly.http.io.OutputBuffer.blockAfterWriteIfNeeded(OutputBuffer.java:999)
```
anymore, but though http2 is enabled I see http1.1 in FF 70:

![image](https://user-images.githubusercontent.com/10200360/70503064-d2575c00-1b22-11ea-93d7-cbc6f063b0b0.png)

**Comment by eduarddrenth** [comment]
chromium shows http1.1 as well

**Comment by rdebusscher** [comment]
Tracked now as `CUSTCOM-22`

**Comment by Budokai100** [comment]
The problem persists in 5.194:

[2019-12-31T10:16:16.236-0500] [Payara 5.194] [WARNING] [jsf.application.resource.unable_to_serve_from_library] [javax.enterprise.resource.webcontainer.jsf.application] [tid: _ThreadID=1186 _ThreadName=http-thread-pool::http-listener-2(24)] [timeMillis: 1577805376236] [levelValue: 900] [[
  JSF1064: Unable to find or serve resource, fonts/rio.eot, from library, primefaces-rio;JSESSIONID=c876a861b38592128c8d2284bc25.]]

the server stops responding after a certain number of requests.

**Comment by oromancer** [comment]
Hello, 

I had this problem too 5.192 + SSL + HTTP/2 with same error.

"The default is true" confused me.
![Default](https://user-images.githubusercontent.com/31956283/72294536-078d2780-3656-11ea-97c3-980139debef5.png)

So I enabled  "HTTP/2 Push Enabled" and problem dissappeared.
Works without problems on **FireFox** 68.0, **Opera** 64.0.3417.92.

**Comment by eduarddrenth** [comment]
Just tested again, with and without push, on 5.194 this: https://github.com/payara/Payara/issues/2625#issuecomment-563894388 is still valid

**Comment by oromancer** [comment]
> Hello,
> 
> I had this problem too 5.192 + SSL + HTTP/2 with same error.
> 
> "The default is true" has confused me.
> 
> ![Default](https://user-images.githubusercontent.com/31956283/72294777-87b38d00-3656-11ea-99b3-876b07c212cd.png)
> 
> So I enabled "HTTP/2 Push Enabled" and problem dissappeared.
> Works without problems on **FireFox** 68.0, **Opera** 64.0.3417.92.

Sorry for that, today problem appeared again.

**Comment by smillidge** [comment]
I've bumped the priority on this.

**Comment by Andrew-Gr** [comment]
Also getting this issue on 5.194 + SSL + HTTP/2. Disabling HTTP/2 fro HTTPS listener resolves the issue.
For us it is not so critical as usually we use nginx as reverse proxy in front of Payara, reporting about this issue for statistics purposes only. Thank you!

**Comment by james16694** [comment]
Moving from 4 to 5.194 build 327 + SSL + HTTP/2 we sometimes see HTML/images loading but not CSS/JS. We also have 2 websites on 5.192 + SSL + HTTP/2 which sometimes display a white screen / or partial loading as 5.194. Both seem related to this issue. We have now disabled HTTP/2 so far all good. We have also seen a class cast exception:

[2020-02-29T23:15:53.175+0000] [Payara 5.194] [WARNING] [] [org.glassfish.grizzly.filterchain.DefaultFilterChain] [tid: _ThreadID=47 _ThreadName=http-thread-pool::http-listener-2(2)] [timeMillis: 1583018153175] [levelValue: 900] [[
  GRIZZLY0013: Exception during FilterChain execution
java.lang.ClassCastException: org.glassfish.grizzly.http2.Http2Request cannot be cast to org.glassfish.grizzly.http.HttpResponsePacket
	at org.glassfish.grizzly.http2.Http2ServerFilter.processOutgoingHttpHeader(Http2ServerFilter.java:960)
	at org.glassfish.grizzly.http2.Http2BaseFilter.handleWrite(Http2BaseFilter.java:729)
	at org.glassfish.grizzly.filterchain.ExecutorResolver$8.execute(ExecutorResolver.java:87)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:260)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:177)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:109)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:88)
	at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:53)
	at org.glassfish.grizzly.filterchain.FilterChainContext.write(FilterChainContext.java:867)
	at org.glassfish.grizzly.filterchain.FilterChainContext.write(FilterChainContext.java:835)
	at org.glassfish.grizzly.http.io.OutputBuffer.flushBuffer(OutputBuffer.java:1068)
	at org.glassfish.grizzly.http.io.OutputBuffer.write(OutputBuffer.java:698)
	at org.apache.catalina.connector.OutputBuffer.writeBytes(OutputBuffer.java:336)
	at org.apache.catalina.connector.OutputBuffer.write(OutputBuffer.java:323)
	at org.apache.catalina.connector.CoyoteOutputStream.write(CoyoteOutputStream.java:154)
	at org.apache.catalina.servlets.DefaultServlet.copy(DefaultServlet.java:2139)
	at org.apache.catalina.servlets.DefaultServlet.serveResource(DefaultServlet.java:1025)
	at org.apache.catalina.servlets.DefaultServlet.doGet(DefaultServlet.java:508)

Thanks

**Comment by MattGill98** [comment]
Hi all,

Thank you all for your input on this issue. This has now spread into several issues, which the work on `CUSTCOM-22` has gone towards resolving. This comment is to provide my summary of each of the issues as far as I understand them.

For each of the issues I've been using the reproducer built from https://bitbucket.org/luminositylabs/payara-2625.git.

# Issue 1

Since these problems are all caused by HTTP/2 push, we advised that turning off HTTP/2 push would fix the issue. Despite this, the following errors could still be seen in the logs:

~~~
 Unable to push resource identified by path [/javax.faces.resource/css/styles.css.action?ln=layout]]]

org.glassfish.grizzly.http2.Http2StreamException streamId=204 errorCode=REFUSED_STREAM
        at org.glassfish.grizzly.http2.Http2Session.openStream(Http2Session.java:1090)
        at org.glassfish.grizzly.http2.Http2ServerFilter.doPush(Http2ServerFilter.java:1017)
        at org.glassfish.grizzly.http2.Http2ServerFilter.handleEvent(Http2ServerFilter.java:603)
~~~

This shows that HTTP/2 push is still, in some capacity, being used. CUSTCOM-22 has now been tagged and will be closed with the linked PR above. This fixes the help text of the admin console, and prevents these errors appearing when HTTP/2 push is disabled, therefore meaning that disabling HTTP/2 push should prevent these errors entirely.

# Issue 2

The second issue is that when reloading the reproducer page repeatedly with push enabled, the following would occasionally happen:

![Capture](https://user-images.githubusercontent.com/12214116/81938011-3c9d8c80-95ec-11ea-8846-7e35cd90fd65.PNG)

A resource intended to be pushed is instead only partially pushed. A thread dump reveals that the server thinks it has finished processing the resource, but the browser is waiting on more. This will need a more complex solution, so `CUSTCOM-277` has been created to track it. This may be related to the errors shown above, but that's not a certainty as sometimes this happens without the errors. I suspect that the errors may be part of the previous connection that couldn't be completed due to a subsequent page reload.

# Issue 3

The third issue is the above error itself. As stated briefly above, this could not be a problem at all if the errors are due to the previous connection being cut by a page reload (essentially the HTTP/2 push version of https://stackoverflow.com/questions/642061/error-commiting-response-java-io-ioexception-broken-pipe-at-sun-nio-ch-filedisp). Once CUSTCOM-277 has been resolved, we can then begin to investigate this properly.

Thank you all for your patience with this issue. If the first issue I outlined is solved by the linked PR, I'll close this issue and reopen issues outlining the remaining HTTP/2 push problems.

**Comment by rdebusscher** [comment]
CUSTCOM-22 is part of release 5.2020.2. More work around HTTP/2 and HTTP Push is planned for next release.

**Comment by AlanRoth** [comment]
Hi, the issue seems to have been resolved in release 5.2020.2 and 5.20. I will be closing this issue, however feel free to raise a new issue if any bugs are spotted in the latest release. Thank you.

**Comment by eduarddrenth** [comment]
Just tested, I can confirm http2 is active and application runs fine. Thanks!

**Comment by lprimak** [comment]
Congratulations everyone! This one is near and dear to my heart.

**Comment by StevenHachel** [comment]
Sorry, but under Payara 5.2022.5 Community I still have the problem as soon as I switch to https and HTTP/2 (not push) is activated. After a short time, resources are simply not reloaded and the display of the pages is incorrect and the behavior is incorrect because the Javascripts sometimes do not work properly.
I have a Primefaces Enterprise application running that doesn't cause any problems under http and Http/2.
Switching to Payara 6 is currently not possible.

**Comment by StevenHachel** [comment]
@marcreichman-pfi No, unfortunately no solution has been found yet. Must still leave this option disabled. The Payara team has not yet come up with a solution that has helped me.

**Comment by smillidge** [comment]
The original issues was closed 4 years ago. If you still see problems I would create a new issue with details.
